### DISCUSSION TITLE
How to setup a 2/3 column Home View?

### DISCUSSION THREAD

**Discussion by ghost** [wwt/SwiftCurrent#167] [Q&A]
I'm trying to get into the secrets of SwiftCurrent, but struggling at several points.

What I'd like to achieve is to set up a home view that is able to switch between a two- and three-column layout, depending on the choice done in the sidebar view.

The problem here is that the sidebar view has `NavigationLink` s with a `WorkflowLauncher`  for each for the destination view. 

What would be necessary is to hand over somehow an indicator, if the selection shall result in a two- or in a three-column layout. 

The `HomeView` itself is just a switch between these two values, with a root `NavigationView` with two or three destination in it.

While this is working in a classic approach, I have no clue how to hand back the selection from the sidebar view to the superior home view. From the documentation I understood that there is no view hierarchy with SwiftCurrent (what's actually what I'd need), but even with the articles, I found no way to do this. 

Would be anyone so kind to give me a hint how this can be achieved?

**Comment by Richard-Gist** [comment]
Trying to understand here.... You have an option between seeing 2 columns or 3 columns.  You have a view that toggles between 2 and 3, and that selection is floated back to HomeView.  HomeView then shows 2 or 3 columns in a NavigationView.  A NavigationView with a ColumnNavigationViewStyle?  Is all that correct?

**Reply by ghost** [reply]
yes, that sounds correct.

**Reply by Richard-Gist** [reply]
Using columns is currently not officially supported.  If you use `.embedInNavigationView()` when creating your `WorkflowLauncher`, it will use the preferred navigation view style (see [implementation in WorkflowLauncher.swift](https://github.com/wwt/SwiftCurrent/blob/bc28d62d6e64b5bb44e88002f542078ac39f2702/Sources/SwiftCurrent_SwiftUI/Views/WorkflowLauncher.swift#L192)) which tries to use the stack style.

HOWEVER, because `WorkflowLauncher` is just a view which itself contains `NavigationLink`s, you can embed the `WorkflowLauncher` inside of a `NavigationView` that you create and set the style to column style.  I think it will then show additional columns as you proceed through the workflow.  Using columns is not currently supported so you may see weird behavior (@Tyler-Keith-Thompson or  @morganzellers can correct me if I'm wrong.)

So your code in `MainView` may look something like:
```swift
var body: some view {
  NavigationView {
    WorkflowLauncher(isLaunched: $isLaunched, args: "String in") {
         thenProceed(with: FirstView.self) {
             thenProceed(with: SecondView.self)
               .presentationType(.navigationLink)
         }.presentationType(.navigationLink)
     }
  }.navigationViewStyle(.columns)
}
```

Hopefully that helps.

**Reply by morganzellers** [reply]
Correct that columns are not currently supported. This looks like it should get you where you're wanting @innoreq - please come back with any more questions 👍🏻

**Comment by Tyler-Keith-Thompson** [comment]
So I read your request completely differently. I'm not seeing that you're using `embedInNavigationView` it seems like you already have your own `NavigationLink` with the destination as a `WorkflowLauncher` and you're just looking for a way to know in your sidebar what selection was chosen. For this I have 2 suggestions:

1. Use `onFinish` and just have it be an output of whatever workflow is in your sidebar. 
```swift
WorkflowLauncher(isLaunched: .constant(true)) {
    thenProceed(with: FirstView.self)
}
.onFinish { print($0) } // This'll give you whatever FirstView outputs when `proceedInWorkflow` is called.
```

2. Use a binding value, much like you would without SwiftCurrent.
```swift
WorkflowLauncher(isLaunched: .constant(true), args: $columnSelection) {
    thenProceed(with: FirstView.self)
}

struct FirstView: View, FlowRepresentable {
    weak var _workflowPointer: AnyFlowRepresentable?
    
    @Binding var columnSelection: <#Type of column selection#>

    init(with columnSelection: Binding<<#Type of column selection#>>) {
        _columnSelection = columnSelection
    }
}
```

Now your binding will be updated just like it would anywhere else.

NOTE: None of this code was run through a compiler, it's just off the top of my head.

If neither answer is helpful try adding a code snippet so we can grok the problem space a bit more.

**Comment by nickkaczmarek** [comment]
@innoreq Did any of this help you out or would you like further assistance?
